src/JitInterface/src/SystemVStructClassificator.cs

```diff
@@ -0,0 +1,648 @@
+                    }
+                }
+                else
+                {
+                    // Ignore static and literal fields as they don't contribute to the layout
+                    do
+                    {
+                        field = fieldEnumerator.Current;
+                        fieldEnumerator.MoveNext();
+                    }
+                    while (field.IsLiteral || field.IsStatic);
+
+                    fieldOffset = field.Offset.AsInt;
+                }
+
+                int normalizedFieldOffset = fieldOffset + startOffsetOfStruct;
+
+                int fieldSize = field.FieldType.GetElementSize().AsInt;
+                Debug.Assert(fieldSize != -1);
+
+                // The field can't span past the end of the struct.
+                if ((normalizedFieldOffset + fieldSize) > helper.structSize)
+                {
+                    Debug.Assert(false, "Invalid struct size. The size of fields and overall size don't agree");
+                    return false;
+                }
+
+                SystemVClassificationType fieldClassificationType;
+                if (typeDesc.IsByReferenceOfT)
+                {
+                    // ByReference<T> is a special type whose single IntPtr field holds a by-ref potentially interior pointer to GC
+                    // memory, so classify its field as such
+                    Debug.Assert(numIntroducedFields == 1);
+                    Debug.Assert(field.FieldType.IsWellKnownType(WellKnownType.IntPtr));
+
+                    fieldClassificationType = SystemVClassificationType.SystemVClassificationTypeIntegerByRef;
+                }
+                else
+                {
+                    fieldClassificationType = TypeDef2SystemVClassification(field.FieldType);
+                }
+
+                if (fieldClassificationType == SystemVClassificationType.SystemVClassificationTypeStruct)
+                {
+                    bool inEmbeddedStructPrev = helper.inEmbeddedStruct;
+                    helper.inEmbeddedStruct = true;
+
+                    bool structRet = false;
+                    structRet = ClassifyEightBytes(field.FieldType, ref helper, normalizedFieldOffset);
+                    
+                    helper.inEmbeddedStruct = inEmbeddedStructPrev;
+
+                    if (!structRet)
+                    {
+                        // If the nested struct says not to enregister, there's no need to continue analyzing at this level. Just return do not enregister.
+                        return false;
+                    }
+
+                    continue;
+                }
+
+                if (fieldClassificationType == SystemVClassificationType.SystemVClassificationTypeTypedReference || 
+                    TypeDef2SystemVClassification(typeDesc) == SystemVClassificationType.SystemVClassificationTypeTypedReference)
+                {
+                    // The TypedReference is a very special type.
```

Comment:
I know it's unrelated to what you're doing but... sigh, at some point we really ought to make TypedReference on CoreCLR same as TypedReference in CoreRT (the first field is `ByReference<T>` instead of `IntPtr` and everything just naturally falls out).

The CoreCLR approach with special casing leaks out _everywhere_.